cpp/src/arrow/compute/kernels/scalar_temporal.cc

```diff
@@ -0,0 +1,614 @@
+  auto func = std::make_shared<ScalarFunction>(name, Arity::Unary(), doc);
+
+  for (auto ty : types) {
+    for (auto unit : AllTimeUnits()) {
+      ArrayKernelExec exec;
+      switch (unit) {
+        case TimeUnit::SECOND: {
+          exec = SimpleUnaryFromOp<Op, std::chrono::seconds>(ty);
+        }
+        case TimeUnit::MILLI: {
+          exec = SimpleUnaryFromOp<Op, std::chrono::milliseconds>(ty);
+        }
+        case TimeUnit::MICRO: {
+          exec = SimpleUnaryFromOp<Op, std::chrono::microseconds>(ty);
+        }
+        case TimeUnit::NANO: {
+          exec = SimpleUnaryFromOp<Op, std::chrono::nanoseconds>(ty);
+        }
+      }
+      auto output_type =
+          struct_({field("iso_year", ty), field("iso_week", ty), field("weekday", ty)});
+      ScalarKernel kernel =
+          ScalarKernel({match::TimestampTypeUnit(unit)}, OutputType(output_type), exec);
+      DCHECK_OK(func->AddKernel(kernel));
+    }
+  }
+  return func;
+}
+
+const FunctionDoc year_doc{"Extract year values", "", {"values"}};
+const FunctionDoc month_doc{"Extract month values", "", {"values"}};
+const FunctionDoc day_doc{"Extract day values", "", {"values"}};
+const FunctionDoc day_of_week_doc{"Extract day of week values", "", {"values"}};
+const FunctionDoc day_of_year_doc{"Extract day of year values", "", {"values"}};
+const FunctionDoc iso_year_doc{"Extract ISO year values", "", {"values"}};
+const FunctionDoc iso_week_doc{"Extract ISO week values", "", {"values"}};
+const FunctionDoc iso_calendar_doc{"Extract ISO calendar values", "", {"values"}};
+const FunctionDoc quarter_doc{"Extract quarter values", "", {"values"}};
+const FunctionDoc hour_doc{"Extract hour values", "", {"values"}};
+const FunctionDoc minute_doc{"Extract minute values", "", {"values"}};
+const FunctionDoc second_doc{"Extract second values", "", {"values"}};
+const FunctionDoc millisecond_doc{"Extract millisecond values", "", {"values"}};
+const FunctionDoc microsecond_doc{"Extract microsecond values", "", {"values"}};
+const FunctionDoc nanosecond_doc{"Extract nanosecond values", "", {"values"}};
+const FunctionDoc subsecond_doc{"Extract subsecond values", "", {"values"}};
+
+void RegisterScalarTemporal(FunctionRegistry* registry) {
+  static std::vector<std::shared_ptr<DataType>> kUnsignedFloatTypes8 = {
+      uint8(), int8(),   uint16(), int16(),   uint32(),
+      int32(), uint64(), int64(),  float32(), float64()};
+  static std::vector<std::shared_ptr<DataType>> kUnsignedIntegerTypes8 = {
+      uint8(), int8(), uint16(), int16(), uint32(), int32(), uint64(), int64()};
+  static std::vector<std::shared_ptr<DataType>> kUnsignedIntegerTypes16 = {
+      uint16(), int16(), uint32(), int32(), uint64(), int64()};
+  static std::vector<std::shared_ptr<DataType>> kUnsignedIntegerTypes32 = {
+      uint32(), int32(), uint64(), int64()};
+  static std::vector<std::shared_ptr<DataType>> kSignedIntegerTypes = {int32(), int64()};
```

Comment:
I don't understand why you're having all these types. The input type is timestamp, right?

Reply:
[See](https://github.com/apache/arrow/pull/10176#discussion_r643454925)